File: src/javascript/lib/string/chars.js
import Kernel from "../kernel";
import Atom from "../atom";
import Integer from "../integer";
import List from "../list";
import * as Core from '../core';
import Tuple from "../tuple";
import { IntegerType, FloatType } from "../protocol";

let Chars = Kernel.defprotocol({
  to_string: function(thing){}
});

Kernel.defimpl(Chars, Core.BitString, {
  to_string: function(thing){
    if(Kernel.is_binary(thing)){
      return thing;
    }

    return thing.toString();
  }
});

Kernel.defimpl(Chars, Symbol, {
  to_string: function(thing){
    if(nil){
      return "";
    }

    return Atom.to_string(thing);
  }
});

Kernel.defimpl(Chars, IntegerType, {
  to_string: function(thing){
    return Integer.to_string(thing);
  }
});

Kernel.defimpl(Chars, FloatType, {
  to_string: function(thing){
    return thing.toString;
  }
});

Kernel.defimpl(Chars, Array, {
  to_string: function(thing){
    return thing.toString();
  }
});

Kernel.defimpl(Chars, Core.Tuple, {
  to_string: function(thing){
    return Tuple.to_string(thing);
  }
});


Kernel.defimpl(Chars, null, {
  to_string: function(thing){
    return thing.toString();
  }
});

export default Chars;

File: lib/elixir_script/prelude/string/chars.ex
defprotocol ElixirScript.String.Chars do
  def to_string(item)
end

defimpl ElixirScript.String.Chars, for: Atom do
  def to_string(nil) do
    ""
  end

  def to_string(atom) do
    Atom.to_string(atom)
  end
end

defimpl ElixirScript.String.Chars, for: BitString do
  def to_string(thing) when is_binary(thing) do
    thing
  end

  def to_string(thing) do
    thing.toString()
  end
end

defimpl ElixirScript.String.Chars, for: List do
  def to_string(list) do
    list.toString()
  end
end

defimpl ElixirScript.String.Chars, for: Tuple do
  def to_string(tuple) do
    tuple.toString()
  end
end

defimpl ElixirScript.String.Chars, for: Integer do
  def to_string(integer) do
    integer.toString()
  end
end

defimpl ElixirScript.String.Chars, for: Float do
  def to_string(float) do
    float.toString()
  end
end
